futures-util/src/future/future/map.rs

```diff
@@ -24,13 +24,26 @@ pin_project! {
 impl<Fut, F> Map<Fut, F> {
     /// Creates a new Map.
     pub(crate) fn new(future: Fut, f: F) -> Self {
-        Self::Incomplete { future, f: Some(f) }
+        Self::Incomplete { future, f }
+    }
+
+    fn take_function(self: Pin<&mut Self>) -> F {
+        // SAFETY The `f` field is not pinned so we can move it out and return it
+        unsafe {
+            match core::mem::replace(self.get_unchecked_mut(), Self::Complete) {
```

Comment:
This moves `future` field.  I think it is not sound, even if you don't access the field (it moves before the drop).
So I think you have to use [code like this](https://github.com/rust-lang/futures-rs/pull/2148/files#diff-1b645f83a773708969f122114e61b5c7e67898aa58c6d701e9d6f0337a3d746aL60-L86). (It's almost equivalent to what the pin-project's [`project_replace`](https://docs.rs/pin-project/1.0.3/pin_project/attr.pin_project.html#project_replace-method) method is [actually using](https://github.com/taiki-e/pin-project/blob/master/examples/project_replace-expanded.rs).)

Reply:
Yep, just realized that and browsed through what project_replace does. It is not as clear cut as I thought it was. Do you think `pin-project-lite` could/should support project_replace as well or is that out of scope? It would have helper in https://github.com/rust-lang/futures-rs/pull/2283 as well, and perhaps in other places in futures-rs also

Reply:
I think it's theoretically can be implemented and it's okay to add it to pin-project-lite, but I'm not sure how easy it is to implement (currently, I don't have the time to investigate it).